doc/source/rllib/rllib-connector.rst

```diff
@@ -0,0 +1,272 @@
+This is done through the ``on_policy_output()`` API call and is useful when your policy is a
+recurrent network, attention network, or auto-regressive model.
+
+
+ActionConnector
+~~~~~~~~~~~~~~~
+
+``ActionConnector`` has a simpler API, which operates on individual actions:
+
+.. code-block:: python
+
+    class ActionConnector(Connector):
+        def __call__(
+            self, ac_data: ActionConnectorDataType
+        ) -> ActionConnectorDataType:
+            ...
+
+        def transform(
+            self, ac_data: ActionConnectorDataType
+        ) -> ActionConnectorDataType:
+            ...
+
+In this case, ``__call__`` and ``transform`` are equivalent. Users may choose to override either
+API to implement an ActionConnector.
+
+Common Data Types
+-----------------
+
+AgentConnectorDataType
+~~~~~~~~~~~~~~~~~~~~~~
+
+Per-agent observation data that goes through an ``AgentConnector`` is in the format of ``AgentConnectorDataType``.
+
+.. literalinclude:: ../../../rllib/utils/typing.py
+   :language: python
+   :start-after: __sphinx_doc_begin_agent_connector_data_type__
+   :end-before: __sphinx_doc_end_agent_connector_data_type__
+
+AgentConnectorsOutput
+~~~~~~~~~~~~~~~~~~~~~
+
+The output from RLlib's default agent connector pipeline is in ``AgentConnectorsOutput`` format.
+
+.. literalinclude:: ../../../rllib/utils/typing.py
+   :language: python
+   :start-after: __sphinx_doc_begin_agent_connector_output__
+   :end-before: __sphinx_doc_end_agent_connector_output__
+
+Note that if ``AgentConnector._is_training`` is False, the ``for_training`` field of ``AgentConnectorsOutput``
+will not be populated. This is for more efficient inference.
+
+.. note::
+    There is a plan to consolidate training episode building into agent connectors, in which case there will
+    not be a need to output the ``for_training`` data piece.
+
+ActionConnectorDataType
+~~~~~~~~~~~~~~~~~~~~~~~
+
+``ActionConnectorDataType`` is the data type ``ActionConnector`` deals with.
+It is basically env and agent IDs plus ``PolicyOutputType``.
+
+.. literalinclude:: ../../../rllib/utils/typing.py
+   :language: python
+   :start-after: __sphinx_doc_begin_action_connector_output__
+   :end-before: __sphinx_doc_end_action_connector_output__
+
+Before, users of RLlib policies would have to come up with the right observation and state inputs
+before they can call a policy. With agent connectors, this task is taken care of automatically.
+
+.. literalinclude:: ../../../rllib/utils/typing.py
+   :language: python
+   :start-after: __sphinx_doc_begin_policy_output_type__
+   :end-before: __sphinx_doc_end_policy_output_type__
+
+
+Advanced Connectors
+-------------------
+
+Lambda Connectors helps turn simple transformation functions into agent or action
+connectors without having users worry about the high level list or non-list APIs.
+Lambda Connectors has separate agent and action versions, for example:
```

Comment:
Same as above, not sure about singular/plural `s` here.

Reply:
done